src/typecheck/mod.rs

```diff
@@ -1856,175 +2306,191 @@ pub fn get_uop_type(
 pub fn get_bop_type(
     state: &mut State,
     op: &BinaryOp,
-) -> Result<(TypeWrapper, TypeWrapper, TypeWrapper), TypecheckError> {
```

Comment:
Types returned by this function shouldn't depend on the environment. So it should return `TypeWrapper`s as before, and the caller would map `Closure::atomic` over the triple. If it's simpler it can return `Closure` by doing it itself, but it shouldn't take an environment as a parameter, and put it into the closure.